### PR TITLE
Add support of RotatE for link prediction.

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*
#980 

*Description of changes:*
Add RotatE score function for link prediction.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/model/edge_decoder.py [hunk 1]
```diff
@@ -1071,6 +1095,403 @@ def get_relembs(self):
...
+    Parameters
+    ----------
+    etypes: list of tuples
+        The canonical edge types of the graph in the format of
+        [(src_ntype1, etype1, dst_ntype1), ...]
+    h_dim: int
+        The input dimension size. It is the dimension for both source and destination
+        node embeddings.
+    gamma: float
+        The gamma value for model initialization and score function. Default: 40.
```
Comment (zhjwy9343) [inline python/graphstorm/model/edge_decoder.py:1128]: This class use `4.0` for the `gamma`. Should we keep all `gamma` be `40.`?

File: python/graphstorm/model/edge_decoder.py [hunk 2]
```diff
@@ -1071,6 +1095,403 @@ def get_relembs(self):
...
+                scores_etype = calc_rotate_pos_score(src_emb,
+                                                     dest_emb,
+                                                     rel_embedding,
+                                                     self.emb_init,
+                                                     self.gamma)
+                scores[canonical_etype] = scores_etype
+
+            return scores
+
+class LinkPredictWeightedRotatEDecoder(LinkPredictRotatEDecoder):
```
Comment (zhjwy9343) [inline python/graphstorm/model/edge_decoder.py:1419]: May add doc string for the new argument `edge_weight_fields`.

File: python/graphstorm/eval/utils.py [hunk 3]
```diff
@@ -213,6 +214,166 @@ def calc_dot_neg_head_score(heads, tails, num_chunks, chunk_size,
...
+
+        .. math:
+
+            d_r(h, t)=\|h\circ r-t\|
+
+        Its score function is defined as:
+
+        .. math:
+
+            gamma - \|h\circ r-t\|^2
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/eval/utils.py:231]: We use \|h\circ r-t\|^2 in DGL-KE paper.

File: python/graphstorm/model/edge_decoder.py [hunk 4]
```diff
@@ -1034,16 +1038,36 @@ def __init__(self,
...
+            .. code:: python
+
+                def init_w_relation(self, gamma):
+                    self._w_relation = nn.Embedding(self.num_rels, self.h_dim)
+
+                    nn.init.uniform_(self._w_relation.weight, -1.0, 1.0)
+
+        Parameters
+        ----------
+        gamma: float
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/model/edge_decoder.py:1067]: RotatE uses it.

[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/model/edge_decoder.py:1067]: Some others may also use it.

File: python/graphstorm/model/edge_decoder.py (hunk 1 above)
Comment (classicsong) [reply python/graphstorm/model/edge_decoder.py:1128]: No. Different decoder (score func) has different gamma value. 40. for RotatE is too large by default.

File: python/graphstorm/eval/utils.py [hunk 5]
```diff
@@ -213,6 +214,172 @@ def calc_dot_neg_head_score(heads, tails, num_chunks, chunk_size,
...
+        ------
+        rotate_score: th.Tensor
+            The RotatE score.
+    """
+    if device is not None:
+        r_emb = r_emb.to(device)
+        heads = heads.to(device)
+        tails = tails.to(device)
+    hidden_dim = heads.shape[1]
+    emb_real = tails[..., :hidden_dim // 2]
```
Comment (Oxfordblue7) [inline python/graphstorm/eval/utils.py:311]: Why using `hidden_dim // 2` here?

File: python/graphstorm/eval/utils.py [hunk 6]
```diff
@@ -213,6 +214,172 @@ def calc_dot_neg_head_score(heads, tails, num_chunks, chunk_size,
...
+        ------
+        rotate_score: th.Tensor
+            The RotatE score.
+    """
+    if device is not None:
+        r_emb = r_emb.to(device)
+        h_emb = h_emb.to(device)
+        t_emb = t_emb.to(device)
+
+    re_head, im_head = th.chunk(h_emb, 2, dim=-1)
```
Comment (Oxfordblue7) [inline python/graphstorm/eval/utils.py:261]:
```suggestion
    real_head, img_head = th.chunk(h_emb, 2, dim=-1)
```
Could you change all `re` to `real` and all `im` to `img` in this function, to keep consistency with other score functions?

File: tests/unit-tests/test_decoder.py [hunk 7]
```diff
@@ -22,16 +22,18 @@
...
                               LinkPredictDistMultDecoder,
-                              LinkPredictWeightedDotDecoder,
-                              LinkPredictWeightedDistMultDecoder,
                               EntityRegression,
                               MLPEFeatEdgeDecoder,
                               LinkPredictContrastiveDotDecoder,
-                              LinkPredictContrastiveDistMultDecoder)
+                              LinkPredictContrastiveDistMultDecoder,
+                              LinkPredictRotatEDecoder,
+                              LinkPredictContrastiveRotatEDecoder)
```
Comment (Oxfordblue7) [inline tests/unit-tests/test_decoder.py:30]: need to test for `LinkPredictWeightedRotatEDecoder `

File: tests/unit-tests/test_decoder.py [hunk 8]
```diff
@@ -22,16 +22,18 @@
...
 from graphstorm.dataloading import (BUILTIN_LP_UNIFORM_NEG_SAMPLER,
                                     BUILTIN_LP_JOINT_NEG_SAMPLER)
-from graphstorm.eval.utils import calc_distmult_pos_score
+from graphstorm.eval.utils import (calc_distmult_pos_score,
+                                   calc_rotate_pos_score)
```
Comment (Oxfordblue7) [inline tests/unit-tests/test_decoder.py:34]: need to test for `calc_distmult_neg_head_score`, `calc_distmult_neg_tail_score`, `calc_rotate_neg_head_score`, and `calc_rotate_neg_tail_score`.

Comment (classicsong) [reply tests/unit-tests/test_decoder.py:34]: They are covered in check_calc_test_scores_rotate_joint_neg and check_calc_test_scores_rotate_uniform_neg

File: python/graphstorm/eval/utils.py (hunk 5 above)
Comment (classicsong) [reply python/graphstorm/eval/utils.py:311]: The node embedding is composed of real part and imaginary part.

File: python/graphstorm/model/edge_decoder.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/model/edge_decoder.py:1068]: As this is a base class, we may remove this argument to let inheritances set their own arguments.

**Review by zhjwy9343 — APPROVED** [review]
LGTM

**Pull request merged**